src/output.rs

```diff
@@ -21,11 +38,92 @@ pub fn print_activation(activated: bool) {
     eprintln!("\x1b[1;34m{} {}.\x1b[0m", word, SHADOWENV);
 }
 
-fn backticks_to_bright_green(err: failure::Error) -> String {
+fn backticks_to_bright_green(err: Error) -> String {
     let re = Regex::new(r"`(.*?)`").unwrap();
     // this is almost certainly not the best way to do this, but this runs at most once per
     // execution so I only care so much.
     let before = format!("{}", err);
     re.replace_all(before.as_ref(), "\x1b[1;32m$1\x1b[1;31m")
         .to_string()
 }
+
+fn check_and_trigger_cooldown(err: &Error) -> Result<bool, Error> {
+    // if no .shadowenv.d, then Err(_) just means no cooldown: always display error.
+    let root = loader::find_root(env::current_dir()?, loader::DEFAULT_RELATIVE_COMPONENT)?
+        .ok_or_else(|| format_err!("no .shadowenv.d"))?;
+
+    let _ = clean_up_stale_errors(&root, Duration::new(300, 0));
+
+    let errindex =
+        cooldown_index(err).ok_or_else(|| format_err!("error not subject to cooldown"))?;
+
+    let errfilepath = err_file(&root, errindex)?;
+
+    match check_cooldown_sentinel(&errfilepath, cooldown()) {
+        Ok(true) => Ok(true),
+        _ => {
+            create_cooldown_sentinel(errfilepath)?;
+            Ok(false)
+        }
+    }
+}
+
+fn cooldown_index(err: &Error) -> Option<u32> {
+    match err.downcast_ref::<trust::NotTrusted>() {
+        Some(_) => Some(0),
+        None => None,
+    }
+}
+
+fn clean_up_stale_errors(root: &PathBuf, timeout: Duration) -> Result<(), Error> {
+    let now = SystemTime::now();
+    if root.is_dir() {
+        for entry in fs::read_dir(root)? {
+            let entry = entry?;
+            if !entry.file_name().to_string_lossy().starts_with(".error-") {
+                continue;
+            }
+            if let Ok(mtime) = entry.metadata().and_then(|md| md.modified()) {
```

Comment:
I'm kind of annoyed that I can't chain `.and_then()` when the chained calls return different error types.